pilot/pkg/serviceregistry/aggregate/controller.go

```diff
@@ -45,28 +45,67 @@ type Controller struct {
 
 // NewController creates a new Aggregate controller
 func NewController() *Controller {
+	log.Infof("><SB> Aggregate Controller NewController was called")
 	return &Controller{
 		registries: make([]Registry, 0),
 	}
 }
 
 // AddRegistry adds registries into the aggregated controller
 func (c *Controller) AddRegistry(registry Registry) {
+	log.Infof("><SB> Aggregate Controller add registry")
 	c.storeLock.Lock()
 	defer c.storeLock.Unlock()
 	c.registries = append(c.registries, registry)
 }
 
+// DeleteRegistry deletes specified registry from the aggregated controller
+func (c *Controller) DeleteRegistry(clusterID string) {
+	c.storeLock.Lock()
+	defer c.storeLock.Unlock()
+	if len(c.registries) == 0 {
+		log.Warnf("Registry list is empty, nothing to delete")
+		return
+	}
+	index, ok := c.GetRegistryIndex(clusterID)
+	if !ok {
+		log.Warnf("Registry is not found in the registries list, nothing to delete")
+		return
+	}
+	c.registries = append(c.registries[:index], c.registries[index+1:]...)
+}
+
+// GetRegistries returns a copy of all registries
+func (c *Controller) GetRegistries() []Registry {
+	c.storeLock.Lock()
+	defer c.storeLock.Unlock()
+
+	registries := make([]Registry, len(c.registries))
+	copy(registries, c.registries)
+	return registries
+}
+
+// GetRegistryIndex returns the index of a registry
+func (c *Controller) GetRegistryIndex(clusterID string) (int, bool) {
+	for i, r := range c.registries {
```

Comment:
Still not happy. Can you instead copy the registries when you add/remove - and return the same slice, without copy, in get ? Copy is rare event, GetRegistry is extremely frequent. Memory allocations/GC do matter.

Reply:
I mean add/remove is rare.